### DISCUSSION TITLE
Update row style from DataTable when pressing a given key

### DISCUSSION THREAD

**Discussion by msempere** [Textualize/textual#2184] [Q&A] [answered]
Hi!
I've been reading the documentation and there's something I don't fully understand how put the pieces together to achieve it. Let's say I have a `DataTable` that gets populated from a CSV file and I'd like when pressing the `a` key to change the style of that row, set all values from all cells to red color, like and _add_ operation, so then, evaluate all those _added_ rows and perform an operation with them.
Is the only way using `update_cell` on each individual column value or is there a more optimal way?

**Comment by TomJGooding** [comment, marked as answer]
> I'd like when pressing the `a` key to change the style of that row, set all values from all cells to red color [...]

[This recent discussion might be helpful](https://github.com/Textualize/textual/discussions/2101). You can apply styles programmatically with rich renderables, but not with CSS as this works at the widget level.

Here's a quick example, though there might be a more efficient way:

```python
from textual.app import App, ComposeResult
from textual.coordinate import Coordinate
from textual.widgets import DataTable, Footer

ROWS = [
    ("lane", "swimmer", "country", "time"),
    (4, "Joseph Schooling", "Singapore", 50.39),
    (2, "Michael Phelps", "United States", 51.14),
    (5, "Chad le Clos", "South Africa", 51.14),
    (6, "László Cseh", "Hungary", 51.14),
    (3, "Li Zhuhao", "China", 51.26),
    (8, "Mehdy Metella", "France", 51.58),
    (7, "Tom Shields", "United States", 51.73),
    (1, "Aleksandr Sadovnikov", "Russia", 51.84),
    (10, "Darren Burns", "Scotland", 51.84),
]


class TableApp(App):
    BINDINGS = [
        ("a", "style_highlighted_row", "Apply styling to highlighted row"),
    ]

    def compose(self) -> ComposeResult:
        yield DataTable()
        yield Footer()


    def on_mount(self) -> None:
        table = self.query_one(DataTable)
        rows = iter(ROWS)
        table.add_columns(*next(rows))
        table.add_rows(rows)

    def action_style_highlighted_row(self) -> None:
        table = self.query_one(DataTable)
        row_idx: int = table.cursor_row
        row_cells = table.get_row_at(row_idx)
        for col_idx, cell_value in enumerate(row_cells):
            cell_coordinate = Coordinate(row_idx, col_idx)
            table.update_cell_at(
                cell_coordinate,
                f"[red]{cell_value}[/red]",
            )


app = TableApp()
if __name__ == "__main__":
    app.run()

```


>  [...] like and add operation, so then, evaluate all those added rows and perform an operation with them.

Sorry but I'm not sure I understand what you mean here - could you provide an example?

[EDIT: Forgot to add the footer!]

**Reply by msempere** [reply]
Ah! that's great, thanks! I was confused by the fact that I could not find anything to perform the row update in one go. But looks like update_cell or update_cell_at is the paved path here.

**Reply by TomJGooding** [reply]
Glad to help, as you can see from the linked discussion I didn't really understand the DataTable myself until recently!

It looks like a `remove_row` method is definitely planned (see #2116), so there's a good chance that a `update_row` method is also somewhere on the roadmap.

**Reply by pawamoy** [reply]
Unfortunately, rows can contain anything, not just strings, so using Rich markup to style a cell casts the cell's value to a string, and we lose information :thinking:

**Reply by pawamoy** [reply]
Also, if I modify the state of the cell value (which is a renderable) without actually changing the value (with `update_cell` or `update_cell_at`), just calling `refresh` is not enough, I need to also increment the internal counter: `table._update_count += 1` before the refresh.

**Reply by TomJGooding** [reply]
I'm not sure what you mean by "modify the state of the cell value [...] without actually changing the value ". Could you give an example?

**Reply by pawamoy** [reply]
Sure, sorry for not being more clear. For example, the first column of each row contains a checkbox: a custom class with a simple `checked` attribute and a `toggle()` method.

```python
class Checkbox:
    def __init__(self):
        self.checked = False

    def __str__(self) -> str:
        return "■" if self.checked else ""

    def __rich__(self) -> str:
        return "[green]■[/]" if self.checked else ""

    def toggle(self):
        self.checked = not self.checked
```

To toggle the checkbox while handling an event, I can therefore do:

```python
table.get_row(row_key)[0].toggle()
```

I modified the state of the cell without actually updating the cell :slightly_smiling_face:

**Reply by TomJGooding** [reply]
Right, because `get_row` just returns a list of values. You haven't actually updated anything in the DataTable.

**Reply by pawamoy** [reply]
Yep. Hence the call to `refresh()`. But without incrementing the internal counter, it doesn't actually refresh.

**Reply by TomJGooding** [reply]
Sorry I'm actually now more confused. Could you provide a MRE showing how you're using your `Checkbox` class in a `DataTable`?

**Reply by pawamoy** [reply]
Sure!

```python
from textual.app import App
from textual.widgets import DataTable, Static, Input


class Checkbox:
    def __init__(self):
        self.checked = False

    def __rich__(self):
        return "[green]■[/]" if self.checked else ""

    def toggle(self):
        self.checked = not self.checked
        return self.checked


class MyApp(App):
    BINDINGS = [
        ("t", "toggle", "Toggle"),
        ("T", "force_toggle", "Toggle"),
    ]

    def compose(self):
        yield Static(
            "Hit lower t to toggle checkbox, then move focus with tab to refresh.\n"
            "Hit upper T to force toggle checkbox, no need to move focus."
        )
        yield DataTable()
        yield Input()

    def on_mount(self):
        table = self.query_one(DataTable)
        table.cursor_type = "row"
        for column in ("?", "One", "Two"):
            table.add_column(column)
        rows = [
            ("Mario", "Luigi"),
            ("Wario", "Waluigi"),
        ]
        table.add_rows((Checkbox(), *row) for row in rows)

    def action_toggle(self):
        table = self.query_one(DataTable)
        key = table.coordinate_to_cell_key(table.cursor_coordinate).row_key
        table.get_row(key)[0].toggle()
        table.refresh()

    def action_force_toggle(self):
        table = self.query_one(DataTable)
        key = table.coordinate_to_cell_key(table.cursor_coordinate).row_key
        table.get_row(key)[0].toggle()
        table._update_count += 1
        table.refresh()


if __name__ == "__main__":
    app = MyApp()
    app.run()
```

**Reply by chbndrhnns** [reply]
>         table._update_count += 1

Modifying an internal counter seems kind of ugly. Is there another approach here? I am also thinking about a feature request but I do not know how to generalize the request a bit
